File: crm/app/internal/ports/segment_repository.go
// Package ports — outbound port for Segment persistence.
// Nothing here may import sqlite, http, or any adapter package.
package ports

import (
	"context"

	"github.com/vantt/data-integration/crm/app/internal/domain"
)

// SegmentRepository persists and queries Segment and SegmentMember records.
type SegmentRepository interface {
	// Insert stores a new segment row.
	Insert(ctx context.Context, s *domain.Segment) error

	// GetByID returns a segment by primary key, or nil if not found.
	GetByID(ctx context.Context, segmentID string) (*domain.Segment, error)

	// Update persists mutable segment fields (name, description, is_dynamic, definition, owner_user_id).
	Update(ctx context.Context, s *domain.Segment) error

	// List returns all segments ordered by created_at DESC.
	List(ctx context.Context) ([]domain.Segment, error)

	// UpsertMember inserts or updates a segment member (idempotent on PK).
	UpsertMember(ctx context.Context, m *domain.SegmentMember) error

	// DeleteMember removes a party from a segment.
	DeleteMember(ctx context.Context, segmentID, partyID string) error

	// ListMembers returns all members of a segment.
	ListMembers(ctx context.Context, segmentID string) ([]domain.SegmentMember, error)

	// DeleteRuleMembers removes all rule-sourced members from a segment.
	// Called before re-evaluating a dynamic segment so stale members are purged.
	DeleteRuleMembers(ctx context.Context, segmentID string) error
}

File: crm/python/domain/ports/segment_repository.py
from __future__ import annotations

from typing import Protocol

from domain.entities.segment import Segment, SegmentMember
from domain.entities.campaign import Campaign, CampaignTarget


class SegmentRepository(Protocol):
    """Outbound port for Segment and SegmentMember persistence."""

    def insert(self, segment: Segment) -> None:
        """Store a new segment row."""
        ...

    def get_by_id(self, segment_id: str) -> Segment | None:
        """Return a segment by primary key, or None if not found."""
        ...

    def update(self, segment: Segment) -> None:
        """Persist mutable segment fields (name, description, is_dynamic,
        definition, owner_user_id)."""
        ...

    def list(self) -> list[Segment]:
        """Return all segments ordered by created_at DESC."""
        ...

    def upsert_member(self, member: SegmentMember) -> None:
        """Insert or update a segment member (idempotent on PK)."""
        ...

    def delete_member(self, segment_id: str, party_id: str) -> None:
        """Remove a party from a segment."""
        ...

    def list_members(self, segment_id: str) -> list[SegmentMember]:
        """Return all members of a segment."""
        ...

    def delete_rule_members(self, segment_id: str) -> None:
        """Remove all rule-sourced members from a segment.
        Called before re-evaluating a dynamic segment to purge stale members."""
        ...


class CampaignRepository(Protocol):
    """Outbound port for Campaign and CampaignTarget persistence."""

    def insert(self, campaign: Campaign) -> None:
        """Store a new campaign row."""
        ...

    def get_by_id(self, campaign_id: str) -> Campaign | None:
        """Return a campaign by primary key, or None if not found."""
        ...

    def update(self, campaign: Campaign) -> None:
        """Persist mutable campaign fields (name, objective, channel,
        segment_id, status, scheduled_at)."""
        ...

    def list(self) -> list[Campaign]:
        """Return all campaigns ordered by created_at DESC."""
        ...

    def upsert_target(self, target: CampaignTarget) -> None:
        """Insert or ignore a campaign target (idempotent on PK)."""
        ...

    def update_target(self, target: CampaignTarget) -> None:
        """Persist mutable target fields (status, assigned_user_id, last_touch_at,
        converted_order_code, converted_revenue_vnd, converted_at)."""
        ...

    def get_target(self, campaign_id: str, party_id: str) -> CampaignTarget | None:
        """Return a single target row, or None if not found."""
        ...

    def list_targets(self, campaign_id: str) -> list[CampaignTarget]:
        """Return all targets for a campaign."""
        ...

    def list_targets_by_campaign_and_status(
        self, campaign_id: str, status: str
    ) -> list[CampaignTarget]:
        """Return targets filtered by status (empty string = all)."""
        ...
